spec/lib/que/locker_spec.rb

```diff
@@ -0,0 +1,126 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+RSpec.describe Que::Locker do
+  subject(:locker) do
+    described_class.new(
+      queue: queue,
+      cursor_expiry: cursor_expiry,
+      metrics: Que::Metrics.new,
+    )
+  end
+
+  let(:queue) { "default" }
+  let(:cursor_expiry) { 0 }
+
+  describe ".with_locked_job" do
+    before { allow(Que).to receive(:execute).and_call_original }
+
+    # Helper to call the with_locked_job method but ensure our block has actually been
+    # called. Without this, it's possible that we'd never run expectations in our block.
+    def with_locked_job(&block)
+      block_called = false
+      locker.with_locked_job do |job|
+        block.call(job)
+        block_called = true
+      end
+
+      raise "did not call job block" unless block_called
+    end
+
+    # Simulates actual working of a job, which is useful to these tests to free up another
+    # job for locking.
+    def expect_to_work(job)
+      with_locked_job do |actual_job|
+        expect(actual_job[:job_id]).to eql(job[:job_id])
+        expect(Que).to receive(:execute).
+          with("SELECT pg_advisory_unlock($1)", [job[:job_id]])
+
+        # Destroy the job to simulate the behaviour of the queue, and allow our lock query
+        # to discover new jobs.
+        QueJob.find(job[:job_id]).destroy!
+      end
+    end
+
+    # Our tests are very concerned with which cursor we use and when
+    def expect_to_lock_with(cursor:)
+      expect(Que).to receive(:execute).with(:lock_job, [queue, cursor])
+    end
+
+    context "with no jobs to lock" do
+      it "scans entire table and calls block with nil job" do
+        expect(Que).to receive(:execute).with(:lock_job, [queue, 0])
+
+        with_locked_job do |job|
+          expect(job).to be_nil
+        end
+      end
+    end
+
+    context "with just one job to lock" do
+      let!(:job_1) { FakeJob.enqueue(1, queue: queue, priority: 1).attrs }
+      let(:cursor_expiry) { 60 }
+
+      # Pretend time isn't moving, as we don't want to confuse this test with cursor
```

Comment:
"Pretend time isn't moving, as we don't want to test time-based cursor expiry here."